refactor: 🏗️ Update C4 diagrams (#405)
seedcase-project/seedcase-sprout: docs/design/c4-models.qmd

---
title: "C4 Models"
---

This section contains the [C4 Models](https://c4model.com/) for Sprout.
The C4 Model is an established visualization approach to describe the
architecture of a software system. It breaks the system down into four
levels of abstraction: Context, Containers, Components, and Code. The
last level is rarely used nor recommended for most software architecture
since it is used to visualize a high degree of detail that most projects

> don't need. Below is the diagrams for the Context, Containers, and

Reviewer: Do we want to capitalise the c's in **C**ontext, **C**ontainers, **C**omponents every time they are written?

Author: Hmm good question. Yea maybe we don't capitalize. The C4 website doesn't.

Author: hmmm, they only capitalize it when combined with "diagram" ("Context diagram")

> Components (which has three diagrams).

Reviewer: ```suggestion
Components (which includes three diagrams).
```

::: panel-tabset
## Context

> The Context model is mainly used to show the users and any external

Reviewer: ```suggestion
The Context model shows the users and any external
```

> The Context model is mainly used to show the users and any external

Reviewer: Why are they called *models* here, *levels of abstraction* in the intro, and elsewhere just *diagrams*? Can this be aligned or are you referring to different things?

Author: Hmm, model is an abstract construct about a real world thing. A diagram is the literal thing it is on the page (the actual image). So you can have a diagram of a model. They aren't the same things. But I see here it should be actually called a diagram at this point.

> systems that interact with Sprout.

Reviewer: ```suggestion
systems that interact with Sprout. This includes the three user roles *data contributor*, *project admin* and *project owner* described in the [users](users.qmd) post.
```

![C4 Context diagram showing very basic overview of Sprout and its
intended users.](images/context.png)

## Container

> The Container model is used to show larger parts of a system, what they

Reviewer: ```suggestion
The Container model shows the larger parts of the system, what they
```

> are responsible for and how they interact with each other. It also shows

Reviewer: ```suggestion
are responsible for, and how they interact with each other. It also shows
```

> technology choices for each Container.

Reviewer: ```suggestion
the technology choices for each Container.
```

![C4 Container diagram showing larger functional parts of Sprout and
their connections.](images/container.png)

## Component

The Component model offers a more detailed view of the system into what
the smaller parts are that make up an individual Container and that
allow it to fulfill its responsibilities. There are three diagrams for

> the Components: the user interface, the server or internal management,

Reviewer: ```suggestion
the Components: the user interface, management,
```
To align with how it's referred to below

and the databases and storage.

> ### User interface
> 
> ![C4 Component diagram showing the parts that make up the user interface
> Containers.](images/component-ui.png)

Reviewer: ```suggestion
### User interface

This diagram shows the pages of the user interface in the user interface container. In addition, it shows which user roles that can access them. Outside the container is the authentication container and the downstream management container. The authentication container is needed to authenticate users and their access. 

![C4 Component diagram showing the parts that make up the user interface
Containers.](images/component-ui.png)
```

Add what the downstream management container is and what it does (bc I'm not sure :).)

Reviewer: Ah! It's just the management components described below? Then maybe remove "downstream" from the name, so it's referred to with the same name both places :)

Reviewer: And adding short descriptions of the two remaining component diagrams would really help with easing the understanding of them :)

### Management

![C4 Component diagram showing the parts that make up the management
Containers.](images/component-management.png)

> ### Databases

Reviewer: ```suggestion
### Databases and storage
```
